### DISCUSSION TITLE
romcal plugins

### DISCUSSION THREAD

**Discussion by tukusejssirs** [romcal/romcal#677] [Ideas]
I want to discuss the way we implement plugins in romcal. ATM I have no specific idea how to implement them. :wink:

IMO user should be able to add multiple plugins at once, and all of them should be considered during the calendar generation.

In the future, I think we should have plugins for the following tasks (some tasks could be done in a single plugin):

- define additional items or update existing ones in the celebration catalogue (`martyrology.ts`):
   - when we add celebrations in plugins, we need to be able to add items to the celebration catalogue;
   - this might be useful for debugging purposes (before merging the changes to romcal);
- define additional calendar (-s):
   - use case: parish calendars with the parish patron, parish church patron, dedications of churches and chapels, patrons of the seconday [mission] churches (as buildings, as well as communities; British term is _chapels of ease_);
   - this might be useful for debugging purposes (before merging the changes to romcal);
- define additional celebrations or modify existing celebrations in a calendar:
   - this might be useful for debugging purposes (before merging the changes to romcal);
- add a way to select which celebration will be celebrated on a particular day (feria vs optional memorials):
   - this would be useful for planning the liturgical celebrations in a particular parish, or even better: define a celebration/formular for a specific Holy Mass (as different Masses on the same date in the same parish might be celebrated Holy Masses with different formular);
   - don’t forget about votive Masses (incl those to saints/blesseds inscribed in the _Martyrologium Romanum_ on that day):
      - OT: this is why I think we should have more persons/celebrations defined in the celebration catalogue (`martyrology.ts`) than only those that are to be used directly by romcal;
- generate a calendar that would include a national calendar and either all or only selected dioceses of that country:
   - this is how ordos do it (at least in Hungary and Slovakia): they list all celebrations celebrated on the national level and if there are different celebrations celebrated on the diocesan level (e.g. dedications or patronages), they are listed as the last celebration of a day;
- define additional locales:
   - this might be useful for debugging purposes (before merging the changes to romcal);
- add additional strings or update existing ones in `.names` locales:
   - when we add celebrations in plugins, we need to be able to localise them;
   - this might be useful for debugging purposes (before merging the changes to romcal).

**Comment by TobiTenno** [comment]
Church buildings that have a community involved are Parish Churches, otherwise it's just a Church (in America)

**Reply by tukusejssirs** [reply]
By _filia_ (_filiálka_ in Slovak and Czech; the word originates from the Latin _filia_ meaning _daughter_) I meant something like a ‘subparish’, i.e. a part of a parish that has a separate church and a community of believes that attend Holy Masses and other services and prayers in that _filial_ church instead of the parish one. That _filial_ church does not have a separate parish priest / pastor / vicar / parish administrator, but the parish priest/administrator of the ‘main’ parish takes cares of the pastoral needs of the believers of the filia.

A parish can have multiples _filias_.

There is a distinction between a _filia_ and two parishes served by one parish priest (which usually takes place for a definite period of time, until another priest is available for serving in one them a parish priest/administrator).

A _filia_ is usually a separate village/town/suburb near the ‘main’ parish village/town.

We could also call it a _branch_ of the parish, although the ‘main’ parish is not considered a branch.

[Here](https://kostolyslovenska.sk/info/slovnik) I found the following definition:

> **filiálka** - chápaná na územnom princípe ako nejaká komunita, ktorá má nejaký počet obyvateľov a vlastné územie, ale nie je natoľko veľká, aby sa mohla stať samostatnou farnosťou (t. j. nemá splnené všetky kritériá na to, aby sa mohla stať farnosťou).

which could be translated like this:

> **filia** - is understood on the territorial principle [or point of view] as a community that has a certain number of inhabitants and its own territory, but is not large enough to become an independent parish (i.e. it does not meet all the criteria for becoming a parish).

**Reply by tukusejssirs** [reply]
Maybe they are called [in American English] _secondary churches_, cf [this](https://www.eriercd.org/bishop/definitions.html). Or mabye _secondary mission churches_, as these in ‘subparishes’, there are Holy Masses celebrated usually once or twice a week (according to the judgement of the parish priest and his duties and free time).

Another [source](https://www.aol.co.uk/erie-county-catholic-parishes-merge-003015427.html?guccounter=1) defines _parish_ and _secondary church_ like this:

> A parish church is the principal church of a parish while secondary churches are one or more other churches in a parish. A secondary mission church has permission for Mass on Sundays and holy days, or on the evenings preceding them.

**Reply by tukusejssirs** [reply]
In British English, the preferred term seems to be _chapels of ease_ (also written with hyphens: _chapels-of-ease_, cf [this](https://heritageportal.buckinghamshire.gov.uk/theme/tbc2) or [this](https://lms.org.uk/blog/rescript-traditional-mass-press-release-latin-mass-society-and-fiuv)).

**Reply by TobiTenno** [reply]
we don't have a separate word for a community of parishes

**Reply by tukusejssirs** [reply]
Well, according to the sources I referenced above, you have. And it is not a _community of [multiple] parishes_ (kind of deanery), but _communty (-ies) within a parish that have a separate church/chapel [from the parish church] to worship God_.

The terms are:

- [chiefly American English] _secondary [mission] church_;
- [British English] _chapel of ease_.

@TobiTenno, don’t you use the term _secondary [mission] church_? Or haven’t you heard it at least? May I know what diocese you belong to? You can share it with me privately if you want.

**Reply by TobiTenno** [reply]
You can say we "have it" but it is not common parlance amongst the faithful. I'm currently living in the Diocese of Kansas City & St Joseph (Dioecesis Kansanopolitanae–Sancti Josephi). The closest things that we have where multiple parishes share a single Priest as pastor are colloquially all in a larger parish community, but each is distinct, neither is secondary or subordinate to another.

Mission churches are an entirely different concept where they are a mission of another particular church or parish, but even where I live, are essentially full parishes with a higher focus on evangelization to the region, not necessarily a different type of parish altogether.

**Reply by emagnier** [reply]
> Church buildings that have a community involved are Parish Churches, otherwise it's just a Church (in America)

This is the same in France

[the remark this answers is not in the document]
**Reply by emagnier** [reply]
There are some good suggestions here, some of which should be easier to implement by splitting and refactoring certain parts of the core Romcal codebase (including the build tools) into different packages within the monorepo (a process that has already started a bit).

For now, I'm less sure about packaging all the proposed ideas as plugins (I think there are simpler ways to make some of them more accessible). First and foremost, the Romcal API needs to be enriched to allow for more advanced customization of the content provided to Romcal when building the calendars.

Currently, one of the only ways to inject custom content is by patching a generated data bundle. This is what I do for some of my Romcal-based projects, but indeed, it's far from being the most user-friendly solution.

So basically, what I gather is that we need to:
- Improve the way calendars are built. For each calendar definition:
  - Allow more flexibility in the calendar inheritance chain (instead of defining only one parent calendar as it is now).
  - Be able to combine different calendar definitions.
- Simplify the process of contributing to/modifying metadata (catalog + locales) and calendar definitions.
  - Provide more tools for creating content (both inside or outside the Romcal repository).
  - Provide more tools to test data integrity.
  - Review certain metadata, such as titles, so they are both more precise and generic (covering a wide range of situations to reduce the need for custom titles for specific calendars).
- Enable the promotion of a list of optional memorials to the rank of memorial, without having to modify an existing calendar definition.

These different topics would benefit from being addressed in separate threads, as this will make it easier to follow and move these various ideas forward. Similarly, for adding support for votive Masses (a new feature that doesn't yet exist in Romcal).
